### PR TITLE
Remove WalkSectors abstraction

### LINKED ISSUE #686: Remove WalkSectors abstraction in miner actor
It's not worth the complexity.

https://github.com/filecoin-project/specs-actors/pull/648#discussion_r455431725

### PR DESCRIPTION
**Pull request by Stebalien**
This patch introduces a new abstraction for walking sectors, but this one is
much simpler and reusable.

This patch also splits up sector expiration rescheduling into a state and a
deadline function (and pushes the partition invariants down into the partition
itself).

fixes #686

### CODE REVIEW

**Comment by Stebalien** [comment]
I suggest reviewing commit by commit.

File: actors/builtin/miner/deadline_state.go [hunk 1]
```diff
@@ -469,40 +469,40 @@ func (dl *Deadline) TerminateSectors(
...
 	var partition Partition
-	for _, partIdx := range partitionIdxs {
-		partSectors := partitionSectors[partIdx]
+	if err := partitionSectors.ForEach(func(partIdx uint64, sectorNos *abi.BitField) error {
 		if found, err := partitions.Get(partIdx, &partition); err != nil {
-			return NewPowerPairZero(), xerrors.Errorf("failed to load partition %d: %w", partIdx, err)
+			return xerrors.Errorf("failed to load partition %d: %w", partIdx, err)
 		} else if !found {
-			return NewPowerPairZero(), xerrors.Errorf("failed to find partition %d", partIdx)
+			return xerrors.Errorf("failed to find partition %d", partIdx)
```
Comment (anorth) [inline actors/builtin/miner/deadline_state.go:478]: ErrNotFound?

File: actors/builtin/miner/deadline_state.go [hunk 2]
```diff
@@ -1065,3 +1055,56 @@ func (dl *Deadline) RecordProvenSectors(
...
+		return err
+	}
+
+	var rescheduledPartitions []uint64 // track partitions with moved expirations.
+	if err := partitionSectors.ForEach(func(partIdx uint64, sectorNos *abi.BitField) error {
+		var partition Partition
+		if found, err := partitions.Get(partIdx, &partition); err != nil {
+			return xc.ErrIllegalState.Wrapf("failed to load partition %d: %w", partIdx, err)
+		} else if !found {
+			// We failed to find the partition. That's ok, we'll skip it.
```
Comment (anorth) [inline actors/builtin/miner/deadline_state.go:1074]: Why would this be ok? Could you give more context as to a state that could cause this? Is it due to compaction and the best-effort deadline-partition index queue?

File: actors/builtin/miner/deadline_state.go [hunk 3]
```diff
@@ -1065,3 +1055,56 @@ func (dl *Deadline) RecordProvenSectors(
...
+			return nil
+		}
+
+		moved, err := partition.RescheduleExpirations(store, sectors, ssize, quant, expiration, sectorNos)
```
Comment (anorth) [inline actors/builtin/miner/deadline_state.go:1078]: Nit: I've been pushing the `ssize` and `qant` params to come last in every method they appear in.

File: actors/builtin/miner/deadline_state.go [hunk 4]
```diff
@@ -1065,3 +1055,56 @@ func (dl *Deadline) RecordProvenSectors(
...
+	partitions, err := dl.PartitionsArray(store)
+	if err != nil {
+		return err
+	}
+
+	var rescheduledPartitions []uint64 // track partitions with moved expirations.
+	if err := partitionSectors.ForEach(func(partIdx uint64, sectorNos *abi.BitField) error {
+		var partition Partition
+		if found, err := partitions.Get(partIdx, &partition); err != nil {
+			return xc.ErrIllegalState.Wrapf("failed to load partition %d: %w", partIdx, err)
```
Comment (anorth) [inline actors/builtin/miner/deadline_state.go:1072]: Unnecessary exit code wrapping here and other error returns below.

File: actors/builtin/miner/miner_state.go [hunk 5]
```diff
@@ -143,11 +143,6 @@ type SectorOnChainInfo struct {
 }
 
 // Location of a specific sector
```
Comment (anorth) [inline actors/builtin/miner/miner_state.go:145]: Remove comment too

File: actors/builtin/miner/miner_state.go [hunk 6]
```diff
@@ -418,190 +413,47 @@ func (st *State) FindSector(store adt.Store, sno abi.SectorNumber) (uint64, uint
...
 
-// Walks the given sectors, deadline by deadline, partition by partition,
-// skipping missing partitions/sectors.
-func (st *State) WalkSectors(
-	store adt.Store,
-	locations []SectorLocation,
-	beforeDeadlineCb func(dlIdx uint64, dl *Deadline) (update bool, err error),
-	partitionCb func(dl *Deadline, partition *Partition, dlIdx, partIdx uint64, sectors *bitfield.BitField) (update bool, err error),
-	afterDeadlineCb func(dlIdx uint64, dl *Deadline) (update bool, err error),
+// Schedule's each sector to expire at its next deadline end.
```
Comment (anorth) [inline actors/builtin/miner/miner_state.go:416]:
```suggestion
// Schedules each sector to expire at its next deadline end.
```

File: actors/builtin/miner/miner_state.go [hunk 7]
```diff
@@ -418,190 +413,47 @@ func (st *State) FindSector(store adt.Store, sno abi.SectorNumber) (uint64, uint
...
-		var deadlineUpdated bool
-		dl, err := deadlines.LoadDeadline(store, uint64(dlIdx))
+		dl, err := deadlines.LoadDeadline(store, dlIdx)
 		if err != nil {
 			return err
 		}
 
-		deadlineUpdated, err = beforeDeadlineCb(uint64(dlIdx), dl)
-		if err != nil {
+		if err := dl.RescheduleSectorExpirations(store, sectors, ssize, quant, newExpiration, pm); err != nil {
```
Comment (anorth) [inline actors/builtin/miner/miner_state.go:444]: Nit: `ssize` and `quant` last.

File: actors/builtin/miner/partition_state.go [hunk 8]
```diff
@@ -238,17 +238,46 @@ func (p *Partition) RemoveRecoveries(sectorNos *abi.BitField, power PowerPair) (
...
+		return nil, err
+	}
+
+	// Filter out terminated sectors.
+	live, err := bitfield.SubtractBitField(present, p.Terminated)
+	if err != nil {
+		return nil, err
+	}
+
+	// And faulty sectors.
```
Comment (anorth) [inline actors/builtin/miner/partition_state.go:257]:
```suggestion
	// Filter out faulty sectors.
```

("and" is very context sensitive, and I misread it as "add")

File: actors/builtin/miner/sector_map.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+// Count counts the number of partitions & sectors within the map.
+func (dm DeadlineSectorMap) Count() (partitions, sectors uint64, err error) {
+	for dlIdx, pm := range dm { //nolint:nomaprange
+		partCount, sectorCount, err := pm.Count()
+		if err != nil {
+			return 0, 0, xerrors.Errorf("when counting deadline %d: %w", dlIdx, err)
+		}
+		if partCount > math.MaxUint64-partitions {
```
Comment (anorth) [inline actors/builtin/miner/sector_map.go:44]: Go doesn't have overflow-safe arithmetic functions?

It's a great insight to add these checks. We should write our own arithmetic wrappers and do this everywhere.

File: actors/builtin/miner/sector_map.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+// Add records the given sector bitfield at the given deadline/partition index.
+func (dm DeadlineSectorMap) Add(dlIdx, partIdx uint64, sectorNos *bitfield.BitField) error {
+	if dlIdx >= WPoStPeriodDeadlines {
+		return xerrors.Errorf("invalid deadline %d", dlIdx)
+	}
+	dl, ok := dm[dlIdx]
+	if !ok {
+		dl = make(PartitionSectorMap, 1)
```
Comment (anorth) [inline actors/builtin/miner/sector_map.go:64]: The capacity of `1` here isn't informed by knowledge of the eventual size, and may lead to more allocations (but only sometimes less memory). It's probably better to just leave it unspecified.

File: actors/builtin/miner/sector_map_test.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+	"github.com/filecoin-project/go-bitfield"
+	rlepluslazy "github.com/filecoin-project/go-bitfield/rle"
+	"github.com/filecoin-project/specs-actors/actors/builtin/miner"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDeadlineSectorMap(t *testing.T) {
+	dm := make(miner.DeadlineSectorMap)
```
Comment (anorth) [inline actors/builtin/miner/sector_map_test.go:16]: Please also tests counts, iteration etc on an empty map, and map with an empty partition.

File: actors/builtin/miner/deadline_state.go (hunk 2 above)
Comment (Stebalien) [reply actors/builtin/miner/deadline_state.go:1074]: See the new comments.

File: actors/builtin/miner/deadline_state.go (hunk 3 above)
Comment (Stebalien) [reply actors/builtin/miner/deadline_state.go:1078]: Ah, I've been doing the opposite (common args first, important args later). But either way is fine as long as we're consistent.

File: actors/builtin/miner/sector_map.go (hunk 9 above)
Comment (Stebalien) [reply actors/builtin/miner/sector_map.go:44]: Unfortunately, no, they don't. It's not an issue in 99% of cases, but this is one of the few where it is (because someone can pass an arbitrarily large bitfield).

Of course, now that we're restricting bitfield runs to `2**63-1` instead of `2**64`, this may become a non-issue.

File: actors/builtin/miner/sector_map.go (hunk 10 above)
Comment (Stebalien) [reply actors/builtin/miner/sector_map.go:64]: How so? I know I'm going to add _at least_ one element in `dl.Add`, that's why I'm pre-allocating for it.

Looking at the source code, it looks like go always allocates at least one bucket (i.e., if the hint is 0-8, one bucket is allocated regardless).

Comment (anorth) [reply actors/builtin/miner/sector_map.go:64]: You're probably more familiar with internals (tho they might change). But specifying 1 could cause the allocation of fewer buckets than specifying nothing, which would then cause a re-hashing later when it expands.

Comment (Stebalien) [reply actors/builtin/miner/sector_map.go:64]: Well, there are two possible assumptions:

1. Go will always allocate enough space for at least one element.
2. Providing an allocation hint will never cause go to allocate less than it would otherwise.

Either way, we're making an assumption about how go works (and will work) internally. In this case, I'm making the second assumption. I had actually assumed that maps with no hint started at size 0 (which isn't the case), hence the hint of 1.

Comment (Stebalien) [reply actors/builtin/miner/sector_map.go:64]: Note: I'm happy to change this, I just feel like it's clearer to tell the compiler "please make room for at least one element" when we know we'll need that space.

Comment (anorth) [reply actors/builtin/miner/sector_map.go:64]: I had the opposite assumption for 2, but I'm happy to leave it as is.

Comment (Stebalien) [reply actors/builtin/miner/sector_map.go:64]: Eh, you're right. It's better to avoid specifying things we don't actually know: https://github.com/filecoin-project/specs-actors/pull/898.
